jcl/src/java.base/share/classes/java/lang/StringBuffer.java

```diff
@@ -91,7 +91,7 @@ private void decompress(int min) {
 			newValue = new char[currentCapacity];
 		}
 
-		String.decompress(value, 0, newValue, 0, currentLength);
+		String.decompress(value, 0, newValue, 0, currentCapacity);
```

Comment:
Should this not be `newValue.length` instead? Same question for `StringBuilder`.

Reply:
Note: doing so renders the local `currentLength` unused, so it should be removed.

Reply:
> Should this not be newValue.length instead? Same question for StringBuilder.

`currentCapacity` is used here since it's the max amount of characters to be copied from the old array. newValue.length` can be greater than the size of the old array.

Reply:
The more I look at this, the more I'm inclined to believe the old code was correct.
Please provide an example usage pattern that is problematic.

Perhaps the problem is with `insert(int index, char[] chars, int start, int length)` which doesn't appear to handle the case where a suffix in the buffer must move.

Reply:
Here is an failing code snippet:
```Java
// run with -XX:+CompactStrings
char ch = 256 // any past 255 will do to trigger decompression
StringBuilder sb = new StringBuilder("this is a test");
StringBuilder sb2 = sb.insert(0, ch); // also works with sb.insert(0, String.valueOf(ch)) or sb.insert(0, new char[] {ch}
String str = String.valueOf(ch) + "this is a test";
System.out.println(sb2.toString().equals(str));
```
the `insert` methods all assume that data moved beyond the current length of the buffer/builder will be available, though `decompress` rewrites the array with only the data up to the length.

Reply:
While this may work, I think it does more work than necessary. The capacity of a buffer may be much larger than its current length, in which case `decompress()` will operate on data that it need not consider.

I suggest that the argument supplied to `decompress()` should represent the end of the meaningful data in the array. More specifically:
* each call from `append()` should pass the current length
* calls from `insert()` should pass the expected new length (which is appropriate after a call to `move()`)